x/rollapp/keeper/rollapp.go

```diff
@@ -245,3 +245,109 @@ func (k Keeper) IsRollappStarted(ctx sdk.Context, rollappId string) bool {
 	_, found := k.GetLatestStateInfoIndex(ctx, rollappId)
 	return found
 }
+
+func (k Keeper) MarkRollappAsVulnerable(ctx sdk.Context, rollappId string) error {
+	return k.FreezeRollapp(ctx, rollappId)
+}
+
+// FreezeRollapp marks the rollapp as frozen and reverts all pending states.
+func (k Keeper) FreezeRollapp(ctx sdk.Context, rollappID string) error {
+	rollapp, found := k.GetRollapp(ctx, rollappID)
+	if !found {
+		return gerrc.ErrNotFound
+	}
+
+	rollapp.Frozen = true
+
+	k.RevertPendingStates(ctx, rollappID)
+
+	if rollapp.ChannelId != "" {
+		clientID, _, err := k.channelKeeper.GetChannelClientState(ctx, "transfer", rollapp.ChannelId)
+		if err != nil {
+			return fmt.Errorf("get channel client state: %w", err)
+		}
+
+		err = k.freezeClientState(ctx, clientID)
+		if err != nil {
+			return fmt.Errorf("freeze client state: %w", err)
+		}
+	}
+
+	k.SetRollapp(ctx, rollapp)
+	return nil
+}
+
+func (k Keeper) verifyClientID(ctx sdk.Context, rollappID, clientID string) error {
+	rollapp, found := k.GetRollapp(ctx, rollappID)
+	if !found {
+		return gerrc.ErrNotFound
+	}
+
+	var (
+		emptyRollappChannelID = rollapp.ChannelId == ""
+		emptyClientID         = clientID == ""
+	)
+
+	switch {
+	// both channelID and clientID are empty
+	case emptyRollappChannelID && emptyClientID:
+		return nil // everything is fine, expected scenario
+
+	// channelID is empty while clientID is not
+	case emptyRollappChannelID:
+		return fmt.Errorf("rollapp does not have a channel: rollapp '%s'", rollappID)
+
+	// clientID is empty while channelID is not
+	case emptyClientID:
+		return fmt.Errorf("empty clientID while the rollapp channelID is not empty")
+
+	// both channelID and clientID are not empty
+	default:
+		// extract rollapp channelID
+		extractedClientId, _, err := k.channelKeeper.GetChannelClientState(ctx, "transfer", rollapp.ChannelId)
+		if err != nil {
+			return fmt.Errorf("get channel client state: %w", err)
+		}
+		// compare it with the passed clientID
+		if extractedClientId != clientID {
+			return fmt.Errorf("clientID does not match the one in the rollapp: clientID %s, rollapp clientID %s", clientID, extractedClientId)
+		}
+		return nil
+	}
+}
+
+func (k Keeper) FilterRollapps(ctx sdk.Context, f func(types.Rollapp) bool) []types.Rollapp {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.RollappKeyPrefix))
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close() // nolint: errcheck
+
+	var result []types.Rollapp
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.Rollapp
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		if f(val) {
+			result = append(result, val)
+		}
+	}
+	return result
+}
```

Comment:
There are already some methods to iterate over rollapps by various index
can you integrate more with that?

Because here it will read every rollapp anyway
so it makes more sense to me to just use generic filter function with a slice of rollapps
unless you concerned about oom

But something here is not right, given that we already do work to index rollapps